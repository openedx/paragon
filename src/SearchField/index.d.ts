import React from 'react';
import { SearchFieldAdvancedProps } from './SearchFieldAdvanced';
import { SearchFieldClearButtonProps } from './SearchFieldClearButton';
import { SearchFieldInputProps } from './SearchFieldInput';
import { SearchFieldLabelProps } from './SearchFieldLabel';
import { SearchFieldSubmitButtonProps } from './SearchFieldSubmitButton';

export interface SearchFieldProps {
  onSubmit: (value: string) => void;
  label?: string | React.ReactNode;
  className?: string;
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
  onChange?: (value: string) => void;
  onClear?: () => void;
  onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void;
  placeholder?: string;
  screenReaderText?: {
    label: string | React.ReactNode;
    submitButton: string | React.ReactNode;
    clearButton?: string | React.ReactNode;
  };
  value?: string;
  icons?: {
    submit: React.ReactNode;
    clear?: React.ReactNode;
  };
  formAriaLabel?: string;
  inputProps?: Record<string, any>;
  variant?: 'light' | 'dark';
  disabled?: boolean;
  submitButtonLocation?: 'internal' | 'external';
  buttonText?: string | React.ReactNode;
}

declare const SearchField: React.FC<SearchFieldProps> & {
  Advanced: React.FC<SearchFieldAdvancedProps>;
  Label: React.FC<SearchFieldLabelProps>;
  Input: React.FC<SearchFieldInputProps>;
  ClearButton: React.FC<SearchFieldClearButtonProps>;
  SubmitButton: React.FC<SearchFieldSubmitButtonProps>;
};

export default SearchField;
