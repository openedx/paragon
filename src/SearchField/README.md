---
title: 'SearchField'
type: 'component'
components:
- SearchField
- SearchFieldAdvanced
- SearchFieldClearButton
- SearchFieldInput
- SearchFieldLabel
- SearchFieldSubmitButton
categories:
- Forms
- Navigation
status: 'Stable'
designStatus: 'Needs Review'
devStatus: 'Done'
notes:
---

Search allows users to quickly find content. The ``Search field`` is made up of the ``Text field`` component and an optional ``Button`` component.

## Basic Usage

```jsx live
<SearchField
  onSubmit={value => console.log(`search submitted: ${value}`)}
/>
```

## With an initial value

```jsx live
<SearchField
  onSubmit={value => console.log(`search submitted: ${value}`)}
  value="foobar"
/>
```

## With a placeholder

```jsx live
<SearchField
  onSubmit={value => console.log(`search submitted: ${value}`)}
  placeholder="foobar"
/>
```

## With callbacks

```jsx live
<SearchField
  onSubmit={value => console.log(`search submitted: ${value}`)}
  onChange={value => console.log(`value changed: ${value}`)}
  onFocus={event => console.log(`input focused: ${event}`)}
  onBlur={event => console.log(`input blurred: ${event}`)}
  onClear={() => console.log('search cleared')}
/>
```

## With a custom label

```jsx live
<SearchField
  onSubmit={value => console.log(`search submitted: ${value}`)}
  label="Search:"
/>
```

## With custom screenreader text

```jsx live
<SearchField
  onSubmit={value => console.log(`search submitted: ${value}`)}
  screenReaderText={{
    label: 'buscar',
    clearButton: 'borrar búsqueda',
    submitButton: 'enviar búsqueda',
  }}
/>
```

## With the submit button outside the input

```jsx live
<SearchField
  submitButtonLocation="external"
  onSubmit={value => console.log(`search submitted: ${value}`)}
/>
```

## Advanced Usage

For needs that deviate from the basic usage above, use `<SearchField.Advanced />`. The `children` elements must contain the `SearchField.Label` and `SearchField.Input` components at a minimum.

## With a custom label

```jsx live
<SearchField.Advanced
  onSubmit={value => console.log(`search submitted: ${value}`)}
>
  <SearchField.Label>Search:</SearchField.Label>
  <SearchField.Input />
  <SearchField.ClearButton />
  <SearchField.SubmitButton />
</SearchField.Advanced>
```

## With an initial value

```jsx live
<SearchField.Advanced
  onSubmit={value => console.log(`search submitted: ${value}`)}
  value="foobar"
>
  <SearchField.Label>Search:</SearchField.Label>
  <SearchField.Input />
  <SearchField.ClearButton />
  <SearchField.SubmitButton />
</SearchField.Advanced>
```

## With a placeholder

```jsx live
<SearchField.Advanced
  onSubmit={value => console.log(`search submitted: ${value}`)}
>
  <SearchField.Label>Search:</SearchField.Label>
  <SearchField.Input placeholder="foobar" />
  <SearchField.ClearButton />
  <SearchField.SubmitButton />
</SearchField.Advanced>
```

## With no clear button

```jsx live
<SearchField.Advanced
  onSubmit={value => console.log(`search submitted: ${value}`)}
>
  <SearchField.Label />
  <SearchField.Input />
  <SearchField.SubmitButton />
</SearchField.Advanced>
```

## With no submit or clear buttons

```jsx live
<SearchField.Advanced
  onSubmit={value => console.log(`search submitted: ${value}`)}
>
  <SearchField.Label />
  <SearchField.Input />
</SearchField.Advanced>
```

## Advanced usage with the submit button outside the input

Use class `pgn__searchfield_wrapper` to group input elements apart from the submit button.

```jsx live
<SearchField.Advanced
  onSubmit={value => console.log(`search submitted: ${value}`)}
  submitButtonLocation="external"
>
  <div className="pgn__searchfield_wrapper">
    <SearchField.Label/>
    <SearchField.Input/>
  </div>
  <SearchField.SubmitButton submitButtonLocation="external" />
</SearchField.Advanced>
```
