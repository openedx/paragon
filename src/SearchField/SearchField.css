/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:20 GMT
 */

:root {
  --pgn-spacing-search-field-margin-button: .5rem;
  --pgn-size-search-field-border-radius: 0;
  --pgn-size-search-field-border-width-focus: .3125rem;
  --pgn-size-search-field-border-width-base: .0625rem;
  --pgn-size-search-field-search-input-height: calc(var(--pgn-typography-form-input-line-height-base) * 1em + var(--pgn-spacing-form-input-padding-y-base) * 2);
}
