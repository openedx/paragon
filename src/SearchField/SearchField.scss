@import "variables";

.pgn__searchfield {
  transition: $input-transition;
  border: $search-border-width solid $search-border-color;

  .btn:focus-visible {
    outline: none;
    position: relative;
    transition: $input-transition;

    &::after {
      content: "";
      border: $search-border-focus-width double $search-border-focus-color;
      position: absolute;
      width: 100%;
      height: 100%;
    }
  }

  &.disabled,
  &:disabled {
    opacity: $search-disabled-opacity;
    pointer-events: none;
  }

  &.has-focus:not(.pgn__searchfield--external) {
    position: relative;

    &::after {
      content: "";
      border: $search-border-focus-width double $search-border-focus-color;
      position: absolute;
      width: 100%;
      height: 100%;
    }

    .pgn__searchfield_wrapper {
      box-shadow: 0 0 0 $search-border-width $search-border-color-interaction;
    }
  }

  &:hover,
  &:active {
    border-color: $search-border-color-interaction;
  }

  .form-control {
    border: none;
    border-radius: $search-border-radius;
    height: $input-height-search;

    &:focus {
      box-shadow: none;
    }

    // clear browser 'x' from Internet Explorer
    &::-ms-clear,
    &::-ms-reveal {
      display: none;
      width: 0;
      height: 0;
    }

    // clear browser 'x' from Chrome
    &::-webkit-search-decoration,
    &::-webkit-search-cancel-button,
    &::-webkit-search-results-button,
    &::-webkit-search-results-decoration {
      display: none;
    }

    .form-control:hover,
    .form-control:focus,
    .form-control:active {
      border: none;
    }
  }

  label.has-label-text {
    padding-inline-start: .75rem;
  }

  &.pgn__searchfield--external {
    border: none;

    .btn-primary {
      background: map-get($search-btn-variants, "light");
    }

    .btn-brand {
      background: map-get($search-btn-variants, "dark");
    }

    &.has-focus {
      box-shadow: none;

      .pgn__searchfield_wrapper {
        position: relative;

        &::after {
          content: "";
          border: $search-border-focus-width double $search-border-focus-color;
          position: absolute;
          width: 100%;
          height: 100%;
        }
      }

      .btn-primary {
        background: map-get($search-btn-variants, "light");
      }

      .btn-brand {
        background: map-get($search-btn-variants, "dark");
      }
    }
  }

  .pgn__searchfield_wrapper {
    display: flex;
    align-items: center;
    width: 100%;
    border: $search-border-width solid $search-border-color;

    &:hover,
    &:active {
      border-color: $search-border-color-interaction;
    }
  }
}

.pgn__searchfield__button.btn[type="submit"] {
  border-radius: 0;
  margin-inline-start: $search-button-margin;
}
