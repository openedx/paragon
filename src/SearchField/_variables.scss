$search-btn-variants: (
  "light":  $primary-500,
  "dark":   $brand-500,
);
$search-border-radius:                0 !default;
$search-line-height:                  1.5rem !default;
$search-border-color:                 $gray-500 !default;
$search-border-color-interaction:     $black !default;
$search-border-width:                 .0625rem !default;
$search-border-width-interaction:     .125rem !default;
$search-disabled-opacity:             .3 !default;
$search-button-margin:                .5rem !default;
$input-height-search:                 calc(#{$input-line-height * 1em} + #{$input-padding-y * 2}) !default;
$search-border-focus-color:           $black !default;
$search-border-focus-width:           .3125rem !default;
