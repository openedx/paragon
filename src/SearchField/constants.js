export const STYLE_VARIANTS = [
  'light',
  'dark',
];

export const BUTTON_LOCATION_VARIANTS = [
  'internal',
  'external',
];
