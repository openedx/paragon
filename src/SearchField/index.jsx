import React, { useCallback } from 'react';
import PropTypes from 'prop-types';

import { Search, Close } from '../../icons';
import SearchFieldAdvanced, { SearchFieldContext } from './SearchFieldAdvanced';
import SearchFieldLabel from './SearchFieldLabel';
import SearchFieldInput from './SearchFieldInput';
import SearchFieldClearButton from './SearchFieldClearButton';
import SearchFieldSubmitButton from './SearchFieldSubmitButton';

export const SEARCH_FIELD_SCREEN_READER_TEXT_LABEL = 'search';
export const SEARCH_FIELD_SCREEN_READER_TEXT_SUBMIT_BUTTON = 'submit search';
export const SEARCH_FIELD_SCREEN_READER_TEXT_CLEAR_BUTTON = 'clear search';
export const SEARCH_FIELD_BUTTON_TEXT = 'search';

const STYLE_VARIANTS = [
  'light',
  'dark',
];

const BUTTON_LOCATION_VARIANTS = [
  'internal',
  'external',
];

function SearchField(props) {
  const {
    label,
    placeholder,
    inputProps,
    variant,
    submitButtonLocation,
    buttonText,
    ...others
  } = props;

  const Wrapper = useCallback(
    (wrapperProps) => (submitButtonLocation === 'external'
      ? <div className="pgn__searchfield_wrapper">{wrapperProps.children}</div>
      : wrapperProps.children),
    [submitButtonLocation],
  );

  return (
    <SearchField.Advanced {...others} submitButtonLocation={submitButtonLocation}>
      <Wrapper>
        <SearchField.Label>{label}</SearchField.Label>
        <SearchField.Input placeholder={placeholder} {...inputProps} />
        <SearchField.ClearButton />
      </Wrapper>
      <SearchField.SubmitButton
        variant={variant}
        submitButtonLocation={submitButtonLocation}
        buttonText={buttonText}
      />
    </SearchField.Advanced>
  );
}

SearchField.propTypes = {
  /**
   * Specifies a callback function for when the `SearchField` is submitted by the user. For example:
   *
   *```jsx
   * <SearchField onSubmit={(value) => { console.log(value); } />
   * ```
   */
  onSubmit: PropTypes.func.isRequired,
  /** Specifies the label to use for the input field (e.g., for i18n translations). */
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  /** Specifies a custom class name. */
  className: PropTypes.string,
  /**
   * Specifies a callback function for when the user loses focus in the
   * `SearchField` component. The default is an empty function. For example:
   *
   * ```jsx
   * <SearchField onBlur={event => console.log(event)} />
   * ```
   */
  onBlur: PropTypes.func,
  /**
   * Specifies a callback function for when the value in `SearchField`
   * is changed by the user. The default is an empty function. For example:
   *
   * ```jsx
   * <SearchField onChange={value => console.log(value)} />
   * ```
   */
  onChange: PropTypes.func,
  /**
   * Specifies a callback function for when the value in `SearchField`
   * is cleared by the user. The default is an empty function. For example:
   *
   * ```jsx
   * <SearchField onClear={() => console.log('search cleared')} />
   * ```
   */
  onClear: PropTypes.func,
  /**
   * Specifies a callback function for when the user focuses in the `SearchField`
   * component. The default is an empty function. For example:
   *
   * ```jsx
   * <SearchField onFocus={event => console.log(event)} />
   * ```
   */
  onFocus: PropTypes.func,
  /** Specifies the placeholder text for the input. */
  placeholder: PropTypes.string,
  /**
   * Specifies the screenreader text for both the clear and submit buttons
   * (e.g., for i18n translations). The default is `{ label: 'search',
   * clearButton: 'clear search', searchButton: 'submit search' }`.
   */
  screenReaderText: PropTypes.shape({
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
    submitButton: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
    clearButton: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  }),
  /** Specifies the initial value for the input. The default is an empty string. */
  value: PropTypes.string,
  /**
   * Specifies the icon element(s) to use for the clear and submit buttons.
   * The default is:
   *
   * ```jsx
   * {
   *   submit: import {Search} from '@openedx/paragon/icons';,
   *   clear: import {Close} from '@openedx/paragon/icons'.
   * }
   * ```
   */
  icons: PropTypes.shape({
    submit: PropTypes.element.isRequired,
    clear: PropTypes.element,
  }),
  /**
   * Specifies the aria-label attribute on the form element.
   * This is useful if you use the `SearchField` component more than once on a page.
   */
  formAriaLabel: PropTypes.string,
  /** Props to be passed to the form input */
  inputProps: PropTypes.shape({}),
  /** The button style variant to use. */
  variant: PropTypes.oneOf(STYLE_VARIANTS),
  /** Specifies whether the `SearchField` is disabled. */
  disabled: PropTypes.bool,
  /** Controls whether the search button is internal as an icon or external as a button. */
  submitButtonLocation: PropTypes.oneOf(BUTTON_LOCATION_VARIANTS),
  /**
   * Specifies a text that is displayed on the button.
   * The `submitButtonLocation` prop should be set to `external`.
   */
  buttonText: PropTypes.string,
};

SearchField.defaultProps = {
  label: undefined,
  placeholder: undefined,
  className: undefined,
  formAriaLabel: undefined,
  value: '',
  screenReaderText: {
    label: SEARCH_FIELD_SCREEN_READER_TEXT_LABEL,
    submitButton: SEARCH_FIELD_SCREEN_READER_TEXT_SUBMIT_BUTTON,
    clearButton: SEARCH_FIELD_SCREEN_READER_TEXT_CLEAR_BUTTON,
  },
  icons: {
    clear: Close,
    submit: Search,
  },
  onBlur: () => {},
  onChange: () => {},
  onFocus: () => {},
  onClear: () => {},
  inputProps: {},
  variant: 'light',
  disabled: false,
  submitButtonLocation: 'internal',
  buttonText: SEARCH_FIELD_BUTTON_TEXT,
};

SearchField.Advanced = SearchFieldAdvanced;
SearchField.Label = SearchFieldLabel;
SearchField.Input = SearchFieldInput;
SearchField.ClearButton = SearchFieldClearButton;
SearchField.SubmitButton = SearchFieldSubmitButton;
SearchField.Context = SearchFieldContext;

export default SearchField;
