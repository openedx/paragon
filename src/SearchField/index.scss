@import "variables";

.pgn__searchfield {
  transition: $input-transition;
  border: $search-border-width solid $search-border-color;

  .pgn__searchfield-form {
    display: flex;
    align-items: center;
    width: 100%;
    background-color: $search-form-background-color;
  }

  .btn:focus-visible {
    outline: none;
    position: relative;
    transition: $input-transition;

    &::after {
      content: "";
      border: $search-border-focus-width double $search-border-focus-color;
      position: absolute;
      width: 100%;
      height: 100%;
    }
  }

  &.disabled,
  &:disabled {
    opacity: $search-disabled-opacity;
    pointer-events: none;
  }

  &.has-focus:not(.pgn__searchfield--external) {
    position: relative;

    &::after {
      content: "";
      border: $search-border-focus-width double $search-border-focus-color;
      position: absolute;
      width: 100%;
      height: 100%;
    }

    .pgn__searchfield_wrapper {
      box-shadow: 0 0 0 $search-border-width $search-border-color-interaction;
    }
  }

  &:hover,
  &:active {
    border-color: $search-border-color-interaction;
  }

  .form-control {
    border: none;
    border-radius: $search-border-radius;
    height: $input-height-search;

    &:focus {
      box-shadow: none;
    }

    // clear browser 'x' from Internet Explorer
    &::-ms-clear,
    &::-ms-reveal {
      display: none;
      width: 0;
      height: 0;
    }

    // clear browser 'x' from Chrome
    &::-webkit-search-decoration,
    &::-webkit-search-cancel-button,
    &::-webkit-search-results-button,
    &::-webkit-search-results-decoration {
      display: none;
    }

    .form-control:hover,
    .form-control:focus,
    .form-control:active {
      border: none;
    }
  }

  label.has-label-text {
    padding-inline-start: .75rem;
  }

  &.pgn__searchfield--external {
    border: none;

    &.has-focus {
      box-shadow: none;

      .pgn__searchfield_wrapper {
        position: relative;

        &::after {
          content: "";
          border: $search-border-focus-width double $search-border-focus-color;
          position: absolute;
          width: 100%;
          height: 100%;
        }
      }
    }
  }

  .pgn__searchfield_wrapper {
    display: flex;
    align-items: center;
    width: 100%;
    border: $search-border-width solid $search-border-color;

    &:hover,
    &:active {
      border-color: $search-border-color-interaction;
    }
  }
}

.pgn__searchfield__button.btn[type="submit"] {
  border-radius: 0;
  margin-inline-start: $search-button-margin;
}

.pgn__searchfield__iconbutton-submit,
.pgn__searchfield__iconbutton-reset {
  flex-shrink: 0;
  margin-inline-end: map-get($spacers, 1);
}
