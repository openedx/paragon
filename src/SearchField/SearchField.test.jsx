import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SearchField from '.';

const BUTTON_LOCATION_VARIANTS = [
  'internal',
  'external',
];

const baseProps = {
  onSubmit: () => {},
};

describe('<SearchField /> with basic usage', () => {
  it('should match the snapshot', () => {
    const { container } = render(<SearchField {...baseProps} />);
    expect(container).toMatchSnapshot();
  });

  it('renders SearchField.Advanced component`', () => {
    render(<SearchField {...baseProps} data-testid="advanced-component" />);
    const advancedComponent = screen.getByTestId('advanced-component');
    expect(advancedComponent).toBeInTheDocument();
  });

  it('should pass correct props to `SearchField.Label`', () => {
    const label = 'foobar';
    let props = { ...baseProps, label };
    const { rerender } = render(<SearchField {...props} label={label} />);
    const labelElement = screen.getByLabelText(label);
    expect(labelElement).toBeInTheDocument();

    props = {
      ...baseProps,
      screenReaderText: { label, submitButton: 'submit foobar' },
    };
    rerender(<SearchField {...props} />);
    const srOnlyLabelElement = screen.getByText(label);
    expect(srOnlyLabelElement).toBeInTheDocument();
    expect(srOnlyLabelElement).toHaveClass('sr-only');
  });

  it('should pass correct props to `SearchField.Input`', () => {
    const placeholder = 'foobar';
    const inputTestId = 'foo';
    const props = { ...baseProps, placeholder, inputProps: { 'data-testid': inputTestId } };
    render(<SearchField {...props} />);
    const inputElement = screen.getByTestId(inputTestId);
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveAttribute('placeholder', placeholder);
  });

  it('should use passed in initial `value` prop', () => {
    const value = 'foobar';
    const props = { ...baseProps, value };
    render(<SearchField {...props} />);
    const inputElement = screen.getByRole('searchbox');
    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toHaveValue(value);
  });

  it('should use passed in `screenReaderText` prop', async () => {
    const screenReaderText = {
      label: 'buscar',
      submitButton: 'enviar búsqueda',
      clearButton: 'borrar búsqueda',
    };
    const props = { ...baseProps, screenReaderText };
    render(<SearchField {...props} />);
    const input = screen.getByRole('searchbox', { target: 'submit' });
    const submitLabel = screen.getByLabelText(screenReaderText.label);
    expect(submitLabel).toBeInTheDocument();
    const submitButton = screen.getByRole('button', { name: screenReaderText.submitButton, type: 'submit' });
    expect(submitButton).toBeInTheDocument();
    await userEvent.type(input, 'foobar');
    const resetButton = screen.getByRole('button', { name: screenReaderText.clearButton, type: 'reset' });
    expect(resetButton).toBeInTheDocument();
  });

  it('should add div if `submitButtonLocation` is passed', () => {
    const { container } = render(<SearchField {...baseProps} />);
    expect(container.querySelector('.pgn__searchfield_wrapper')).toBeNull();
    const { container: containerExternal } = render(<SearchField {...baseProps} submitButtonLocation="external" />);
    expect(containerExternal.querySelector('.pgn__searchfield_wrapper')).toBeInTheDocument();
  });

  describe('should fire', () => {
    it('focus handler', () => {
      const spy = jest.fn();
      const props = { ...baseProps, onFocus: spy };
      render(<SearchField {...props} />);
      const inputElement = screen.getByRole('searchbox');
      inputElement.focus();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('blur handler', async () => {
      const spy = jest.fn();
      const props = { ...baseProps, onBlur: spy };
      render(<SearchField {...props} />);
      const inputElement = screen.getByRole('searchbox');
      inputElement.focus();
      await userEvent.tab();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('change handler', async () => {
      const spy = jest.fn();
      const props = { ...baseProps, onChange: spy };
      const inputText = 'foobar';
      render(<SearchField {...props} />);
      const inputElement = screen.getByRole('searchbox');
      await userEvent.type(inputElement, inputText);
      expect(spy).toHaveBeenCalledTimes(inputText.length);
    });

    it('clear handler', async () => {
      const spy = jest.fn();
      const props = { ...baseProps, onClear: spy };
      render(<SearchField {...props} />);
      const inputElement = screen.getByRole('searchbox');
      await userEvent.type(inputElement, 'foobar');

      const resetButton = screen.getByRole('button', { type: 'reset' });
      await userEvent.click(resetButton);

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('submit handler on submit button click', async () => {
      const spy = jest.fn();
      const props = { ...baseProps, onSubmit: spy, submitButtonLocation: BUTTON_LOCATION_VARIANTS[1] };
      render(<SearchField {...props} />);
      const inputElement = screen.getByRole('searchbox');
      const submitButton = screen.getByRole('button', { type: 'submit' });
      await userEvent.type(inputElement, 'foobar');
      await userEvent.click(submitButton);
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith('foobar');
    });
  });

  describe('clear button', () => {
    it('should be visible with input value', async () => {
      const props = { ...baseProps };
      render(<SearchField {...props} />);
      const inputElement = screen.getByRole('searchbox');
      expect(screen.queryByRole('button', { name: 'clear search', type: 'reset' })).toBeNull();
      await userEvent.type(inputElement, 'foobar');
      expect(screen.getByRole('button', { type: 'reset' })).toBeInTheDocument();
    });

    it('should clear input value when clicked', async () => {
      const props = { ...baseProps };
      render(<SearchField {...props} />);
      const inputElement = screen.getByRole('searchbox', { target: 'submit' });
      await userEvent.type(inputElement, 'foobar');
      expect(inputElement).toHaveValue('foobar');
      const clearButton = screen.getByRole('button', { type: 'reset' });
      await userEvent.click(clearButton);
      expect(inputElement).toHaveValue('');
    });
  });

  describe('advanced usage', () => {
    it('should pass props to the clear button', async () => {
      const buttonProps = { variant: 'inline' };
      render(
        <SearchField.Advanced {...baseProps}>
          <SearchField.Input />
          <SearchField.ClearButton {...buttonProps} />
        </SearchField.Advanced>,
      );
      const inputElement = screen.getByRole('searchbox');
      await userEvent.type(inputElement, 'foobar');
      const buttonClear = screen.getByRole('button', { type: 'reset', variant: buttonProps.variant });
      expect(buttonClear).toHaveClass(`btn-icon-${buttonProps.variant}`);
    });

    it('should pass props to the label', () => {
      const labelProps = { variant: 'inline' };
      render(
        <SearchField.Advanced {...baseProps}>
          <SearchField.Label {...labelProps}>Labeled</SearchField.Label>
        </SearchField.Advanced>,
      );
      const label = screen.getByText('Labeled');
      expect(label).toHaveAttribute('variant', 'inline');
    });

    it('should pass props to the submit button', () => {
      const buttonText = 'Some test text';
      const buttonProps = {
        submitButtonLocation: 'external',
        buttonText,
      };
      render(
        <SearchField.Advanced {...baseProps}>
          <SearchField.SubmitButton {...buttonProps} />
        </SearchField.Advanced>,
      );
      const submitButton = screen.getByText(buttonText);
      expect(submitButton).toBeInTheDocument();
      expect(submitButton).toHaveClass('pgn__searchfield__button');
    });

    it('should pass variant to the submit button', () => {
      const buttonText = 'Some test text';
      const buttonProps = {
        submitButtonLocation: 'external',
        buttonText,
      };
      render(
        <SearchField.Advanced {...baseProps}>
          <SearchField.SubmitButton {...buttonProps} variant="dark" />
        </SearchField.Advanced>,
      );
      const submitButton = screen.getByText(buttonText);
      expect(submitButton).toHaveClass('btn-brand');
    });
  });
});
