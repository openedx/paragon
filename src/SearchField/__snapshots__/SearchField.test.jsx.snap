// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchField /> with basic usage should match the snapshot 1`] = `"<div class=\\"pgn__searchfield d-flex\\"><form role=\\"search\\" class=\\"d-flex align-items-center w-100\\"><label for=\\"pgn-searchfield-input-1\\" class=\\"m-0\\"><span class=\\"sr-only\\">search</span></label><input type=\\"text\\" class=\\"form-control\\" role=\\"searchbox\\" id=\\"pgn-searchfield-input-1\\" name=\\"searchfield-input\\" value=\\"\\"/><button type=\\"submit\\" class=\\"btn\\"><span class=\\"pgn__icon\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" focusable=\\"false\\" aria-hidden=\\"true\\"><path d=\\"M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\" fill=\\"currentColor\\"></path></svg></span><span class=\\"sr-only\\">submit search</span></button></form></div>"`;
